frame/crowdloan-rewards/src/lib.rs

```diff
@@ -0,0 +1,325 @@
+			reward_account: T::AccountId,
+			relay_account: T::RelayChainAccountId,
+			proof: MultiSignature,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+			ensure!(
+				verify_relay(
+					T::Prefix::get(),
+					reward_account.clone(),
+					relay_account.clone(),
+					proof
+				),
+				Error::<T>::InvalidProof
+			);
+			let remote_account = RemoteAccount::RelayChain(relay_account);
+			let claimed = Self::do_claim(remote_account.clone(), &reward_account)?;
+			Self::deposit_event(Event::Claimed { remote_account, reward_account, amount: claimed });
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn do_claim(
+			remote_account: RemoteAccountOf<T>,
+			reward_account: &T::AccountId,
+		) -> Result<T::Balance, DispatchError> {
+			ensure!(VestingBlockStart::<T>::exists(), Error::<T>::NotInitialized);
+			Rewards::<T>::try_mutate(remote_account, |reward| {
+				reward
+					.as_mut()
+					.map(|reward| {
+						let upfront_payment = T::InitialPayment::get().mul_floor(reward.total);
+						let should_have_claimed = {
+							let current_block = frame_system::Pallet::<T>::block_number();
+							// Current point in time
+							let vesting_point =
+								current_block.saturating_sub(VestingBlockStart::<T>::get());
+							if vesting_point >= reward.vesting_period {
+								// If the user is claiming when the period is over, he should
+								// probably have already claimed everything.
+								reward.total
+							} else {
+								let vesting_period = VestingBlockStart::<T>::get()
+									.safe_add(&reward.vesting_period)?;
+								let partition = T::VestingPartition::get();
+								// Current window, rounded to previous window.
+								let vesting_window = vesting_point - (vesting_point % partition);
+								// The user should have claimed the upfront payment + the vested
+								// amount until this window point.
+								let vested_reward = reward.total - upfront_payment;
+								upfront_payment +
+									(vested_reward
+										.saturating_mul(T::Convert::convert(vesting_window)) /
+										T::Convert::convert(vesting_period))
+							}
+						};
+						let available_to_claim = should_have_claimed - reward.claimed;
+						ensure!(
+							available_to_claim > T::Balance::zero(),
+							Error::<T>::NothingToClaim
+						);
+						T::Currency::mint_into(reward_account, available_to_claim)?;
+						(*reward).claimed += available_to_claim;
+						ClaimedRewards::<T>::mutate(|x| *x += available_to_claim);
+						Ok(available_to_claim)
+					})
+					.unwrap_or_else(|| Err(Error::<T>::NotAContributor.into()))
+			})
+		}
+	}
+
+	pub fn verify_relay<AccountId: Encode, RelayChainAccountId: Into<AccountId32>>(
+		prefix: &[u8],
+		reward_account: AccountId,
+		relay_account: RelayChainAccountId,
+		proof: MultiSignature,
+	) -> bool {
+		let wrapped_prefix: &[u8] = b"<Bytes>";
+		let wrapped_postfix: &[u8] = b"</Bytes>";
+		let mut msg = wrapped_prefix.to_vec();
+		msg.append(&mut prefix.to_vec());
```

Comment:
may be use some thing like <version:u8>:<signature_type:u8>:<msg>? like in libp2p

Reply:
I don't really understand what you mean here, this prefix/postfix are the one appended by the polkadotjs framework when signing.

Reply:
i see. that is polkadot stuff. thanks.